src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,141 @@
+#include "partial_hash_index_impl.hpp"
+#include "storage/segment_iterate.hpp"
+
+namespace opossum {
+
+template <typename DataType>
+PartialHashIndexImpl<DataType>::PartialHashIndexImpl(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id)
+    : BasePartialHashIndexImpl() {
+  add(chunks_to_index, column_id);
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::add(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id) {
+  size_t size_before = _indexed_chunk_ids.size();
+  for (const auto& chunk : chunks_to_index) {
+    // do not allow multiple indexing of one chunk
+    if (_indexed_chunk_ids.contains(chunk.first)) continue;
+
+    _indexed_chunk_ids.insert(chunk.first);
+    auto indexed_segment = chunk.second->get_segment(column_id);
+    segment_iterate<DataType>(*indexed_segment, [&](const auto& position) {
+      auto row_id = RowID{chunk.first, position.chunk_offset()};
+      if (position.is_null()) {
+        if (!_null_values.contains(true)) {
+          _null_values[true] = std::vector<RowID>();
+        }
+        _null_values[true].push_back(row_id);
+      } else {
+        if (!_map.contains(position.value())) {
+          _map[position.value()] = std::vector<RowID>();
+        }
+        _map[position.value()].push_back(row_id);
+      }
+    });
+  }
+
+  return _indexed_chunk_ids.size() - size_before;
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::remove(const std::vector<ChunkID>& chunks_to_remove) {
+  size_t size_before = _indexed_chunk_ids.size();
+  for (const auto& chunk_id : chunks_to_remove) {
+    if (!_indexed_chunk_ids.contains(chunk_id)) continue;
+
+    _indexed_chunk_ids.erase(chunk_id);
+    auto map_iter = _map.begin();
+    while (map_iter != _map.end()) {
+      auto& values = _map.at(map_iter->first);
+      values.erase(std::remove_if(values.begin(), values.end(),
+                                  [chunk_id](RowID& row_id) { return row_id.chunk_id == chunk_id; }),
+                   values.end());
+      if (values.empty()) {
+        map_iter = _map.erase(map_iter);
+      } else {
+        ++map_iter;
+      }
+    }
+    auto& nulls = _null_values[true];
+    nulls.erase(
+        std::remove_if(nulls.begin(), nulls.end(), [chunk_id](RowID& row_id) { return row_id.chunk_id == chunk_id; }),
+        nulls.end());
+  }
+
+  return size_before - _indexed_chunk_ids.size();
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::IteratorPair PartialHashIndexImpl<DataType>::equals(
```

Comment:
The naming of the function should be improved IMO: `equals` sounds like an equality comparison to me, which returns a boolean value.